### PR TITLE
Provide an ArrayDef

### PR DESCRIPTION
**Pull request by vallsv**
Hi,

Here is the state of my work to create an `ArrayDef`.

I did implement multidim array for fun.

It is close to something reviewable but i have some extra proposals.

I would like to provide extra arguments for `serialize` and `deserialize`.
Basically providing a `buffer` and an `offset`. `serialize` with a buffer with not return the result but update the buffer.
This will allow to remove a lot of mem copy i have introduced. If you don't like this idea i still can create a protected `_serialize`.
This will speed up and also remove few more lines of code cause `_de/serialize_element` will not be needed anymore i think. And it's also nice for a public API.

Then I would like to propose to set `StructDef.add(length=None` by default. This is because it is not possible to create an array of one element, but for now it is interpreted as a scalar data. I have added a FIXME about this in the PR, cause changing the default here breaks some tests. So i prefer to ask before reworking that stuff.

What do you think.

### CODE REVIEW

**Comment by midstar** [comment]
> Hi,
> 
> Here is the state of my work to create an `ArrayDef`.
> 
> I did implement multidim array for fun.
> 
> It is close to something reviewable but i have some extra proposals.
> 
> I would like to provide extra arguments for `serialize` and `deserialize`.
> Basically providing a `buffer` and an `offset`. `serialize` with a buffer with not return the result but update the buffer.
> This will allow to remove a lot of mem copy i have introduced. If you don't like this idea i still can create a protected `_serialize`.
> This will speed up and also remove few more lines of code cause `_de/serialize_element` will not be needed anymore i think. And it's also nice for a public API.
> 
> Then I would like to propose to set `StructDef.add(length=None` by default. This is because it is not possible to create an array of one element, but for now it is interpreted as a scalar data. I have added a FIXME about this in the PR, cause changing the default here breaks some tests. So i prefer to ask before reworking that stuff.
> 
> What do you think.

I like all your suggestions. Go for it :-)

**Comment by vallsv** [comment]
Hi, i have updated the code.

I finally changed my mind, but it's up to you in the end.

- It was not a good idea to remove the `serialize_element`, so i did apply this pattern to the `ArrayDef`
- And i have added a `shape` attribute to `add` for the arrays, but you still have `length` which could be convenient for strings.

So what this creates an array of 5x5 of strings of 10 chars
```
s = StructDef()
s.add("utf-8", "array", 10, shape=(5, 5))
```

I have also introduced the following thing to create array def, but i don't know if it's a nice thing.
```
s = StructDef()
s.add("int8", 'int')
array_of_struct = s[10]
ndim_of_struct = s[10][5][2]  # In C you have to use the reverse [2][5][10] so it could be a problem

s2 = StructDef()
s2.add(s, "ndim_of_struct", shape=(2, 5, 10))  # And here you use the C order
```

Here is a kind of change log:
- Create an `ArrayDef` supporting multi-dimensions
- Rework `StructDef.__str__` to display multi-dimension arrays
- An `ArrayDef` can be created from a base type using indexing `__getitem__`
- Add a `shape` argument to `StructDef.add` 
- Lazy loading of instances for `_ListInstance`
- `serialize/deserialize` methods now supports `buffer` and `offset`

Could you please take a look at the expected changes by black?
On my side it looks like it wants to change the whole indentation of the doc of the `add` method. And i don't understand why.

**Comment by midstar** [comment]
> Hi, i have updated the code.
> 
> I finally changed my mind, but it's up to you in the end.
> 
> * It was not a good idea to remove the `serialize_element`, so i did apply this pattern to the `ArrayDef`
> * And i have added a `shape` attribute to `add` for the arrays, but you still have `length` which could be convenient for strings.
> 
> So what this creates an array of 5x5 of strings of 10 chars
> 
> ```
> s = StructDef()
> s.add("utf-8", "array", 10, shape=(5, 5))
> ```
> 
> I have also introduced the following thing to create array def, but i don't know if it's a nice thing.
> 
> ```
> s = StructDef()
> s.add("int8", 'int')
> array_of_struct = s[10]
> ndim_of_struct = s[10][5][2]  # In C you have to use the reverse [2][5][10] so it could be a problem
> 
> s2 = StructDef()
> s2.add(s, "ndim_of_struct", shape=(2, 5, 10))  # And here you use the C order
> ```
> 
> Here is a kind of change log:
> 
> * Create an `ArrayDef` supporting multi-dimensions
> * Rework `StructDef.__str__` to display multi-dimension arrays
> * An `ArrayDef` can be created from a base type using indexing `__getitem__`
> * Add a `shape` argument to `StructDef.add`
> * Lazy loading of instances for `_ListInstance`
> * `serialize/deserialize` methods now supports `buffer` and `offset`
> 
> Could you please take a look at the expected changes by black?
> On my side it looks like it wants to change the whole indentation of the doc of the `add` method. And i don't understand why.

Sorry, I accidentally closed the PR but its reopened again. I have briefly look at you changes and I think it looks great. I like the idea of using length parameter for strings only and the shape parameter for arrays/matrixes. I posted a comment on the StructDef add method to fix the indentation error in the document. 

It will be some work to update the cparser and that file, to be honest, is a little bit "hacky" so it might be tricky. Tell me if you would like to get some help with that.

File: pycstruct/pycstruct.py [hunk 1]
```diff
@@ -316,8 +512,13 @@ def add(self, datatype, name, length=1, byteorder="", same_level=False):
         :param name: Name of element. Needs to be unique.
         :type name: str
         :param length: Number of elements. If > 1 this is an array/list of
-                       elements with equal size. Default is 1.
+-                      elements with equal size. Default is 1. This should only
```
Comment (midstar) [inline pycstruct/pycstruct.py]: This line starts with -, thats why the doc generation gives warnings and the doc has strange indentation.

Comment (vallsv) [reply pycstruct/pycstruct.py]: Oups... Thanks for the check.

**Comment by vallsv** [comment]
Here is with the cparse implementation. It was finally not so tricky.

**Comment by vallsv** [comment]
Maybe the cov is failing cause there is not so much unsupported ctypes?

**Comment by midstar** [comment]
> Maybe the cov is failing cause there is not so much unsupported ctypes?

You can check which lines are covered by running:

```
coverage run --source=./pycstruct -m unittest discover -s tests/
coverage html
```
Report is here htmlcov/index.html

**Comment by vallsv** [comment]
Yes, sure, but i mean the uncovered lines are failing/skiped branches.

That's painful test to write, and i don't know if this branches are still useful.

I did an extra commit to show it to you.

**Comment by vallsv** [comment]
Oh, just found that few tests are slipped. I have added `-v` to the test runner.

**Comment by vallsv** [comment]
So the problem is part of the CI in the master: `castxml is not installed`

**Comment by vallsv** [comment]
I also did check the diff between master and this PR coverage.
This thing is not covered anymore.

```
!                 except Exception as exception:
!                     logger.warning(
!                         """%s has a member %s could not be handled:
!  - %s
!  - Composite type will be ignored.""",
!                         dict_output["name"],
!                         member["name"],
!                         str(exception.args[0]),
!                     )
!                     dict_output["supported"] = False
!                     break
```
It was covered by `test_xml_parse_special_cases`

Do you think it is still needed? Or should i add a test here. The test case here is not obvious for me.

**Comment by vallsv** [comment]
Ohh... it was covered by
```
struct non_supported_member {
    int matrix[10][10];
};
```
Which is now supported.

Do you think there is still something which is not supported? Else i can clean up this try/except.

**Comment by midstar** [comment]
> Ohh... it was covered by
> 
> ```
> struct non_supported_member {
>     int matrix[10][10];
> };
> ```
> 
> Which is now supported.
> 
> Do you think there is still something which is not supported? Else i can clean up this try/except.

I dont think we have any unsupported types now. You can remove it.

Regarding castxml is not installed, this is expected. That is why I have added "pre generated" xml files in the repository.

**Comment by vallsv** [comment]
Done

**Comment by midstar** [comment]
Your changes looks great. However could you also update the documentation:

- doc/source/overview.rst
- https://github.com/midstar/pycstruct/blob/master/doc/source/limitations.rst

Great work!

**Comment by vallsv** [comment]
Here is an update of the doc. Let me know if it is enough.

**Comment by midstar** [comment]
Everything looks great now, but could you just update one thing more? 

If we merge this pull request it will fail on master since the version number 0.9.1 is already available in pypi.org.

So, to bump the version number you need to update two files:

- appveyor.yml - version on first row
- setup.py - version argument of setup function on 10th row

Suggest we call this version 0.10.0

**Comment by vallsv** [comment]
That's done.

File: setup.py [hunk 2]
```diff
@@ -4,33 +4,35 @@
...
-      'License :: OSI Approved :: MIT License',
-      'Programming Language :: Python :: 3',
-      'Programming Language :: Python :: 3.4',
-      'Programming Language :: Python :: 3.5',
-      'Programming Language :: Python :: 3.6',
-      'Programming Language :: Python :: 3.7'
-      ])
+setup(
+    name="pycstruct",
+    version="0.9.1",
```
Comment (midstar) [inline setup.py]: Change of ' to " made buildtest.py fail. Just update following row in buildtest.py:
`            setup_version = line.split("'")[1]`

Comment (vallsv) [reply setup.py]: Here we are

**Review by midstar — APPROVED** [review]

**Pull request merged**